File: ios/Impermanence/Impermanence/TimeFormatting.swift
import Foundation

enum TimeFormatting {
    private static let twelveHourFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.amSymbol = "AM"
        formatter.pmSymbol = "PM"
        formatter.dateFormat = "h:mm a"
        return formatter
    }()

    private static let twentyFourHourFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "HH:mm"
        return formatter
    }()

    static func formattedTime(from date: Date, use24HourClock: Bool) -> String {
        let formatter = use24HourClock ? twentyFourHourFormatter : twelveHourFormatter
        return formatter.string(from: date)
    }

    static func formattedDuration(from interval: TimeInterval) -> String {
        let totalSeconds = max(0, Int(interval.rounded(.down)))
        let hours = totalSeconds / 3600
        let minutes = (totalSeconds % 3600) / 60
        let seconds = totalSeconds % 60
        return String(format: "%02d:%02d:%02d", hours, minutes, seconds)
    }
}

File: android/app/src/main/java/com/neversink/impermanence/util/TimeFormatting.kt
package com.neversink.impermanence.util

import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.util.Locale

object TimeFormatting {
    private val twelveHourFormatter: DateTimeFormatter = DateTimeFormatter.ofPattern("h:mm a", Locale.US)
    private val twentyFourHourFormatter: DateTimeFormatter = DateTimeFormatter.ofPattern("HH:mm", Locale.US)

    fun formattedTimeFromSeconds(seconds: Int, use24HourClock: Boolean): String {
        val normalized = ((seconds % 86400) + 86400) % 86400
        val time = LocalTime.ofSecondOfDay(normalized.toLong())
        val formatter = if (use24HourClock) twentyFourHourFormatter else twelveHourFormatter
        return time.format(formatter)
    }

    fun formattedDurationMinutes(minutes: Int): String {
        val hours = minutes / 60
        val remainder = minutes % 60
        return when {
            hours == 0 -> "${remainder}m"
            remainder == 0 -> "${hours}h"
            else -> "${hours}h ${remainder}m"
        }
    }
}
